staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go

```diff
@@ -279,8 +281,13 @@ func handleRootHealth(name string, firstTimeHealthy func(), checks ...HealthChec
 
 // adaptCheckToHandler returns an http.HandlerFunc that serves the provided checks.
 func adaptCheckToHandler(c func(r *http.Request) error) http.HandlerFunc {
+	var requestGroup singleflight.Group
+	const checkKey = "check"
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		err := c(r)
+		_, err, _ := requestGroup.Do(checkKey, func() (interface{}, error) {
+			err := c(r)
```

Comment:
this is passing the first request to the health check. if it uses the request's context for cancellation, that could timeout / error the check for other requests with different timeouts

Reply:
Does the request to healthcheck goes through the timeout handlers?

Reply:
yes, but I'm not sure why that's relevant

imagine:
1. request A to /healthz arrives with timeout of 1 second
2. request B to /healthz arrives with timeout of 30 seconds and gets aggregated into the singleflight
3. /healthz takes longer than a second
4. request A times out, failing the /healthz check with a timeout error
5. the timeout response is sent to request B

Reply:
do you mean something like `https://host:6443/healthz?timeout=5?`

AFAIK Unless you have a special handler , you don't have per request timeouts on the server https://github.com/golang/go/issues/54136 , or am I getting this wrong?